Dispatch Test Failing Due to Graph `adj` Comparison for some algorithms in `expanders.py`

The dispatch tests are consistently failing across multiple PRs. The failures occur when checking if the `G.adj == converted_result.adj` condition in the `_convert_and_call_for_tests` while dealing with algorithms that return graph generators. Also, the assertion `assert G.graph == converted_result.graph` passes, but assert `G.adj == converted_result.adj` fails.


ref. https://github.com/networkx/networkx/pull/7443#pullrequestreview-2123949415, https://github.com/networkx/networkx/actions/runs/9557304752/job/26344147953?pr=7503

<details>
<summary> test failure</summary>

```
=================================== FAILURES ===================================
_________ [doctest] networkx.generators.expanders.is_regular_expander __________
356 
357     Returns
358     -------
359     bool
360         Whether the given graph is a regular $(n, d, \lambda)$-expander
361         where $\lambda = 2 \sqrt{d - 1} + \epsilon$.
362 
363     Examples
364     --------
365     >>> G = nx.random_regular_expander_graph(20, 4)
UNEXPECTED EXCEPTION: AssertionError()
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/doctest.py", line 1355, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest networkx.generators.expanders.is_regular_expander[0]>", line 1, in <module>
  File "/home/runner/work/networkx/networkx/networkx/utils/decorators.py", line 789, in func
    return argmap._lazy_compile(__wrapper)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<class 'networkx.utils.decorators.argmap'> compilation 3317", line 4, in argmap_random_regular_expander_graph_3313
    import inspect
            ^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/backends.py", line 715, in __call__
    return self._convert_and_call_for_tests(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/backends.py", line 1266, in _convert_and_call_for_tests
    result = getattr(backend, self.name)(*converted_args, **converted_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/generators/expanders.py", line 459, in random_regular_expander_graph
    G = maybe_regular_expander(
        ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/decorators.py", line 789, in func
    return argmap._lazy_compile(__wrapper)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<class 'networkx.utils.decorators.argmap'> compilation 3322", line 4, in argmap_maybe_regular_expander_3318
    import inspect
            ^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/backends.py", line 715, in __call__
    return self._convert_and_call_for_tests(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/backends.py", line 1472, in _convert_and_call_for_tests
    assert G.adj == converted_result.adj
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
/home/runner/work/networkx/networkx/networkx/generators/expanders.py:365: UnexpectedException
________ [doctest] networkx.generators.expanders.maybe_regular_expander ________
243 
244     The graph is generated by taking $d / 2$ random independent cycles.
245 
246     Joel Friedman proved that in this model the resulting
247     graph is an expander with probability
248     $1 - O(n^{-\tau})$ where $\tau = \lceil (\sqrt{d - 1}) / 2 \rceil - 1$. [1]_
249 
250     Examples
251     --------
252     >>> G = nx.maybe_regular_expander(n=200, d=6, seed=8020)
UNEXPECTED EXCEPTION: AssertionError()
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/doctest.py", line 1355, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest networkx.generators.expanders.maybe_regular_expander[0]>", line 1, in <module>
  File "<class 'networkx.utils.decorators.argmap'> compilation 3322", line 4, in argmap_maybe_regular_expander_3318
    import inspect
            ^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/backends.py", line 715, in __call__
    return self._convert_and_call_for_tests(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/backends.py", line 1472, in _convert_and_call_for_tests
    assert G.adj == converted_result.adj
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
/home/runner/work/networkx/networkx/networkx/generators/expanders.py:252: UnexpectedException
____ [doctest] networkx.generators.expanders.random_regular_expander_graph _____
429       Seed used to set random number generation state. See :ref`Randomness<randomness>`.
430 
431     Raises
432     ------
433     NetworkXError
434         If max_tries is reached
435 
436     Examples
437     --------
438     >>> G = nx.random_regular_expander_graph(20, 4)
UNEXPECTED EXCEPTION: AssertionError()
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/doctest.py", line 1355, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest networkx.generators.expanders.random_regular_expander_graph[0]>", line 1, in <module>
  File "<class 'networkx.utils.decorators.argmap'> compilation 3317", line 4, in argmap_random_regular_expander_graph_3313
    import inspect
            ^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/backends.py", line 715, in __call__
    return self._convert_and_call_for_tests(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/backends.py", line 1266, in _convert_and_call_for_tests
    result = getattr(backend, self.name)(*converted_args, **converted_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/generators/expanders.py", line 459, in random_regular_expander_graph
    G = maybe_regular_expander(
        ^^^^^^^^^^^^^^^^^^^^^^^
  File "<class 'networkx.utils.decorators.argmap'> compilation 3322", line 4, in argmap_maybe_regular_expander_3318
    import inspect
            ^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/backends.py", line 715, in __call__
    return self._convert_and_call_for_tests(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/networkx/networkx/networkx/utils/backends.py", line 1472, in _convert_and_call_for_tests
    assert G.adj == converted_result.adj
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
/home/runner/work/networkx/networkx/networkx/generators/expanders.py:438: UnexpectedException
_______________________ test_maybe_regular_expander[2-7] _______________________

d = 2, n = 7

    @pytest.mark.parametrize("d, n", [(2, 7), (4, 10), (4, 16)])
    def test_maybe_regular_expander(d, n):
        pytest.importorskip("numpy")
>       G = nx.maybe_regular_expander(n, d)

networkx/generators/tests/test_expanders.py:69: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
<class 'networkx.utils.decorators.argmap'> compilation 3322:4: in argmap_maybe_regular_expander_3318
    import inspect
networkx/utils/backends.py:715: in __call__
    return self._convert_and_call_for_tests(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <networkx.utils.backends._dispatchable object at 0x7f484ee18d10>
backend_name = 'nx-loopback', args = (7, 2)
kwargs = {'create_using': None, 'max_tries': 100, 'seed': RandomState(MT19937) at 0x7F484141DD40}

    def _convert_and_call_for_tests(
        self, backend_name, args, kwargs, *, fallback_to_nx=False
    ):
        """Call this dispatchable function with a backend; for use with testing."""
        backend = _load_backend(backend_name)
        if not self._can_backend_run(backend_name, *args, **kwargs):
            if fallback_to_nx or not self.graphs:
                return self.orig_func(*args, **kwargs)
    
            import pytest
    
            msg = f"'{self.name}' not implemented by {backend_name}"
            if hasattr(backend, self.name):
                msg += " with the given arguments"
            pytest.xfail(msg)
    
        from collections.abc import Iterable, Iterator, Mapping
        from copy import copy
        from io import BufferedReader, BytesIO, StringIO, TextIOWrapper
        from itertools import tee
        from random import Random
    
        import numpy as np
        from numpy.random import Generator, RandomState
        from scipy.sparse import sparray
    
        # We sometimes compare the backend result to the original result,
        # so we need two sets of arguments. We tee iterators and copy
        # random state so that they may be used twice.
        if not args:
            args1 = args2 = args
        else:
            args1, args2 = zip(
                *(
                    (arg, copy(arg))
                    if isinstance(
                        arg, BytesIO | StringIO | Random | Generator | RandomState
                    )
                    else tee(arg)
                    if isinstance(arg, Iterator)
                    and not isinstance(arg, BufferedReader | TextIOWrapper)
                    else (arg, arg)
                    for arg in args
                )
            )
        if not kwargs:
            kwargs1 = kwargs2 = kwargs
        else:
            kwargs1, kwargs2 = zip(
                *(
                    ((k, v), (k, copy(v)))
                    if isinstance(
                        v, BytesIO | StringIO | Random | Generator | RandomState
                    )
                    else ((k, (teed := tee(v))[0]), (k, teed[1]))
                    if isinstance(v, Iterator)
                    and not isinstance(v, BufferedReader | TextIOWrapper)
                    else ((k, v), (k, v))
                    for k, v in kwargs.items()
                )
            )
            kwargs1 = dict(kwargs1)
            kwargs2 = dict(kwargs2)
        try:
            converted_args, converted_kwargs = self._convert_arguments(
                backend_name, args1, kwargs1, use_cache=False
            )
            _logger.debug(
                f"using backend '{backend_name}' for call to `{self.name}' "
                f"with args: {converted_args}, kwargs: {converted_kwargs}"
            )
            result = getattr(backend, self.name)(*converted_args, **converted_kwargs)
        except (NotImplementedError, nx.NetworkXNotImplemented) as exc:
            if fallback_to_nx:
                return self.orig_func(*args2, **kwargs2)
            import pytest
    
            pytest.xfail(
                exc.args[0] if exc.args else f"{self.name} raised {type(exc).__name__}"
            )
        # Verify that `self._returns_graph` is correct. This compares the return type
        # to the type expected from `self._returns_graph`. This handles tuple and list
        # return types, but *does not* catch functions that yield graphs.
        if (
            self._returns_graph
            != (
                isinstance(result, nx.Graph)
                or hasattr(result, "__networkx_backend__")
                or isinstance(result, tuple | list)
                and any(
                    isinstance(x, nx.Graph) or hasattr(x, "__networkx_backend__")
                    for x in result
                )
            )
            and not (
                # May return Graph or None
                self.name in {"check_planarity", "check_planarity_recursive"}
                and any(x is None for x in result)
            )
            and not (
                # May return Graph or dict
                self.name in {"held_karp_ascent"}
                and any(isinstance(x, dict) for x in result)
            )
            and self.name
            not in {
                # yields graphs
                "all_triads",
                "general_k_edge_subgraphs",
                # yields graphs or arrays
                "nonisomorphic_trees",
            }
        ):
            raise RuntimeError(f"`returns_graph` is incorrect for {self.name}")
    
        def check_result(val, depth=0):
            if isinstance(val, np.number):
                raise RuntimeError(
                    f"{self.name} returned a numpy scalar {val} ({type(val)}, depth={depth})"
                )
            if isinstance(val, np.ndarray | sparray):
                return
            if isinstance(val, nx.Graph):
                check_result(val._node, depth=depth + 1)
                check_result(val._adj, depth=depth + 1)
                return
            if isinstance(val, Iterator):
                raise NotImplementedError
            if isinstance(val, Iterable) and not isinstance(val, str):
                for x in val:
                    check_result(x, depth=depth + 1)
            if isinstance(val, Mapping):
                for x in val.values():
                    check_result(x, depth=depth + 1)
    
        def check_iterator(it):
            for val in it:
                try:
                    check_result(val)
                except RuntimeError as exc:
                    raise RuntimeError(
                        f"{self.name} returned a numpy scalar {val} ({type(val)})"
                    ) from exc
                yield val
    
        if self.name in {"from_edgelist"}:
            # numpy scalars are explicitly given as values in some tests
            pass
        elif isinstance(result, Iterator):
            result = check_iterator(result)
        else:
            try:
                check_result(result)
            except RuntimeError as exc:
                raise RuntimeError(
                    f"{self.name} returned a numpy scalar {result} ({type(result)})"
                ) from exc
            check_result(result)
    
        if self.name in {
            "edmonds_karp",
            "barycenter",
            "contracted_edge",
            "contracted_nodes",
            "stochastic_graph",
            "relabel_nodes",
            "maximum_branching",
            "incremental_closeness_centrality",
            "minimal_branching",
            "minimum_spanning_arborescence",
            "recursive_simple_cycles",
            "connected_double_edge_swap",
        }:
            # Special-case algorithms that mutate input graphs
            bound = self.__signature__.bind(*converted_args, **converted_kwargs)
            bound.apply_defaults()
            bound2 = self.__signature__.bind(*args2, **kwargs2)
            bound2.apply_defaults()
            if self.name in {
                "minimal_branching",
                "minimum_spanning_arborescence",
                "recursive_simple_cycles",
                "connected_double_edge_swap",
            }:
                G1 = backend.convert_to_nx(bound.arguments["G"])
                G2 = bound2.arguments["G"]
                G2._adj = G1._adj
                if G2.is_directed():
                    G2._pred = G1._pred
                nx._clear_cache(G2)
            elif self.name == "edmonds_karp":
                R1 = backend.convert_to_nx(bound.arguments["residual"])
                R2 = bound2.arguments["residual"]
                if R1 is not None and R2 is not None:
                    for k, v in R1.edges.items():
                        R2.edges[k]["flow"] = v["flow"]
                    R2.graph.update(R1.graph)
                    nx._clear_cache(R2)
            elif self.name == "barycenter" and bound.arguments["attr"] is not None:
                G1 = backend.convert_to_nx(bound.arguments["G"])
                G2 = bound2.arguments["G"]
                attr = bound.arguments["attr"]
                for k, v in G1.nodes.items():
                    G2.nodes[k][attr] = v[attr]
                nx._clear_cache(G2)
            elif (
                self.name in {"contracted_nodes", "contracted_edge"}
                and not bound.arguments["copy"]
            ):
                # Edges and nodes changed; node "contraction" and edge "weight" attrs
                G1 = backend.convert_to_nx(bound.arguments["G"])
                G2 = bound2.arguments["G"]
                G2.__dict__.update(G1.__dict__)
                nx._clear_cache(G2)
            elif self.name == "stochastic_graph" and not bound.arguments["copy"]:
                G1 = backend.convert_to_nx(bound.arguments["G"])
                G2 = bound2.arguments["G"]
                for k, v in G1.edges.items():
                    G2.edges[k]["weight"] = v["weight"]
                nx._clear_cache(G2)
            elif (
                self.name == "relabel_nodes"
                and not bound.arguments["copy"]
                or self.name in {"incremental_closeness_centrality"}
            ):
                G1 = backend.convert_to_nx(bound.arguments["G"])
                G2 = bound2.arguments["G"]
                if G1 is G2:
                    return G2
                G2._node.clear()
                G2._node.update(G1._node)
                G2._adj.clear()
                G2._adj.update(G1._adj)
                if hasattr(G1, "_pred") and hasattr(G2, "_pred"):
                    G2._pred.clear()
                    G2._pred.update(G1._pred)
                if hasattr(G1, "_succ") and hasattr(G2, "_succ"):
                    G2._succ.clear()
                    G2._succ.update(G1._succ)
                nx._clear_cache(G2)
                if self.name == "relabel_nodes":
                    return G2
            return backend.convert_to_nx(result)
    
        converted_result = backend.convert_to_nx(result)
        if isinstance(converted_result, nx.Graph) and self.name not in {
            "boykov_kolmogorov",
            "preflow_push",
            "quotient_graph",
            "shortest_augmenting_path",
            "spectral_graph_forge",
            # We don't handle tempfile.NamedTemporaryFile arguments
            "read_gml",
            "read_graph6",
            "read_sparse6",
            # We don't handle io.BufferedReader or io.TextIOWrapper arguments
            "bipartite_read_edgelist",
            "read_adjlist",
            "read_edgelist",
            "read_graphml",
            "read_multiline_adjlist",
            "read_pajek",
            "from_pydot",
            "pydot_read_dot",
            "agraph_read_dot",
            # graph comparison fails b/c of nan values
            "read_gexf",
        }:
            # For graph return types (e.g. generators), we compare that results are
            # the same between the backend and networkx, then return the original
            # networkx result so the iteration order will be consistent in tests.
            G = self.orig_func(*args2, **kwargs2)
            if not nx.utils.graphs_equal(G, converted_result):
                assert G.number_of_nodes() == converted_result.number_of_nodes()
                assert G.number_of_edges() == converted_result.number_of_edges()
                assert G.graph == converted_result.graph
                assert G.nodes == converted_result.nodes
>               assert G.adj == converted_result.adj
E               AssertionError

networkx/utils/backends.py:1472: AssertionError
______________________ test_maybe_regular_expander[4-10] _______________________

d = 4, n = 10

    @pytest.mark.parametrize("d, n", [(2, 7), (4, 10), (4, 16)])
    def test_maybe_regular_expander(d, n):
        pytest.importorskip("numpy")
>       G = nx.maybe_regular_expander(n, d)

networkx/generators/tests/test_expanders.py:69: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
<class 'networkx.utils.decorators.argmap'> compilation 3322:4: in argmap_maybe_regular_expander_3318
    import inspect
networkx/utils/backends.py:[715](https://github.com/networkx/networkx/actions/runs/9557304752/job/26344147953?pr=7503#step:5:716): in __call__
    return self._convert_and_call_for_tests(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <networkx.utils.backends._dispatchable object at 0x7f484ee18d10>
backend_name = 'nx-loopback', args = (10, 4)
kwargs = {'create_using': None, 'max_tries': 100, 'seed': RandomState(MT19937) at 0x7F484141DD40}

    def _convert_and_call_for_tests(
        self, backend_name, args, kwargs, *, fallback_to_nx=False
    ):
        """Call this dispatchable function with a backend; for use with testing."""
        backend = _load_backend(backend_name)
        if not self._can_backend_run(backend_name, *args, **kwargs):
            if fallback_to_nx or not self.graphs:
                return self.orig_func(*args, **kwargs)
    
            import pytest

.....
.....
.....
```
</details>

Failure summary:
```
=========================== short test summary info ============================
FAILED networkx/generators/expanders.py::networkx.generators.expanders.is_regular_expander
FAILED networkx/generators/expanders.py::networkx.generators.expanders.maybe_regular_expander
FAILED networkx/generators/expanders.py::networkx.generators.expanders.random_regular_expander_graph
FAILED networkx/generators/tests/test_expanders.py::test_maybe_regular_expander[2-7] - AssertionError
FAILED networkx/generators/tests/test_expanders.py::test_maybe_regular_expander[4-10] - AssertionError
FAILED networkx/generators/tests/test_expanders.py::test_maybe_regular_expander[4-16] - AssertionError
FAILED networkx/generators/tests/test_expanders.py::test_random_regular_expander[2-7] - AssertionError
FAILED networkx/generators/tests/test_expanders.py::test_random_regular_expander[4-10] - AssertionError
FAILED networkx/generators/tests/test_expanders.py::test_random_regular_expander[4-16] - AssertionError
FAILED networkx/generators/tests/test_expanders.py::test_random_regular_expander_badinput - AssertionError
= 10 failed, 6185 passed, 59 skipped, 9 xfailed, 6 warnings in 552.70s (0:09:12) =
```

might be related to https://github.com/networkx/networkx/pull/7432

Thank you :)

--- a/networkx/utils/backends.py
+++ b/networkx/utils/backends.py
@@ -1209,7 +1209,7 @@ class _dispatchable:
             pytest.xfail(msg)
 
         from collections.abc import Iterable, Iterator, Mapping
-        from copy import copy
+        from copy import copy, deepcopy
         from io import BufferedReader, BytesIO, StringIO, TextIOWrapper
         from itertools import tee
         from random import Random
@@ -1226,10 +1226,10 @@ class _dispatchable:
         else:
             args1, args2 = zip(
                 *(
-                    (arg, copy(arg))
-                    if isinstance(
-                        arg, BytesIO | StringIO | Random | Generator | RandomState
-                    )
+                    (arg, deepcopy(arg))
+                    if isinstance(arg, RandomState)
+                    else (arg, copy(arg))
+                    if isinstance(arg, BytesIO | StringIO | Random | Generator)
                     else tee(arg)
                     if isinstance(arg, Iterator)
                     and not isinstance(arg, BufferedReader | TextIOWrapper)
@@ -1242,10 +1242,10 @@ class _dispatchable:
         else:
             kwargs1, kwargs2 = zip(
                 *(
-                    ((k, v), (k, copy(v)))
-                    if isinstance(
-                        v, BytesIO | StringIO | Random | Generator | RandomState
-                    )
+                    ((k, v), (k, deepcopy(v)))
+                    if isinstance(v, RandomState)
+                    else ((k, v), (k, copy(v)))
+                    if isinstance(v, BytesIO | StringIO | Random | Generator)
                     else ((k, (teed := tee(v))[0]), (k, teed[1]))
                     if isinstance(v, Iterator)
                     and not isinstance(v, BufferedReader | TextIOWrapper)

